---
title: "Scratch"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Visualization}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```

Original version: 31 October, 2023

# Scratch for visualization testing

Testing markdown integration with gosling tracks, shiny gosling plots are shiny apps which may or may not embed natively in Rmds. Worth experimenting with Rmd and develop ideal plots for alphamissense package

Start by loading the [AlphaMissenseR][] library.

[AlphaMissenseR]: https://mtmorgan.github.io/AlphaMissenseR

```{r setup, message = FALSE}
library(AlphaMissenseR)
```

Visit the summary of available AlphaMissense datasets

```{r am_available}
am_available()
```

This vignette uses the `aa_substitutions` and `hg38` data resources;
make sure that these have been cached locally.

```{r am_data}
am_data("aa_substitutions")
am_data("hg38")
```

# Visualizing sequence tracks

The variant prediction data can also be visualized on a genome browser. Mis-sense point mutations can be tracked alongside pathogenicity on multiple scales by utilizing Gosling, a declarative genomics visualization library.  

```{r}
library(GenomicRanges)

tbl <-
    am_data("hg38") |>
    filter(uniprot_id == "Q1W6H9")

gpos <-
    tbl |>
    to_GPos()
as_GRanges <- as(gpos, "GRanges")
as_GRanges

```

##  Track visualization


#Reference-to-Alternative Allele


```{r, tootip_track}
library(shiny.gosling)
library(shiny)

ui <- fluidPage(
  use_gosling(clear_files = FALSE),
  goslingOutput("gosling_plot")
)


track1_data <- track_data_gr(
    as_GRanges,chromosomeField = "seqnames", genomicFields = c("start", "end"))

ui <- fluidPage(
    use_gosling(clear_files = FALSE),
    goslingOutput("gosling_plot")
)

track_1 <- add_single_track(
    width = 800,
    height = 180,
    data = track1_data,
    mark = "bar",
    x = visual_channel_x(
        field = "start", type = "genomic", axis = "bottom"
    ),
    xe = visual_channel_x(field = "end", type = "genomic"),
    y = visual_channel_y(
        field = "am_pathogenicity", type = "quantitative", axis = "right"
    ),
    color = visual_channel_color(
        field = "am_pathogenicity",
        type = "quantitative"
    ),
        tooltip = visual_channel_tooltips(
      visual_channel_tooltip(field = "REF", type = "nominal",
                             alt = "Reference"),
      visual_channel_tooltip(field = "ALT", type = "nominal",
                             alt = "Alternative / Mutation"),
      visual_channel_tooltip(
        field = "am_pathogenicity",
        type = "quantitative",
        alt = "AM_Pathogenicity Score",
        format = "0.2"
      ) ),
    size = list(value = 5)
)

composed_view <- compose_view(
    layout = "linear",
    xDomain = list(chromosome = "chr2",interval = c(46000, 46400)),
    tracks = track_1
    
)

arranged_view <- arrange_views(
    title = "Test track track : Q1W6H9 ",
    subtitle = "Bar example with tooltips",
    views = composed_view
)

server <- function(input, output, session) {
    output$gosling_plot <- renderGosling({
        gosling(
            component_id = "component_1",
            arranged_view
        )
    })
}

shiny::shinyApp(ui, server)
```


```{r, text_track}
library(shiny.gosling)
library(shiny)

ui <- fluidPage(
  use_gosling(clear_files = FALSE),
  goslingOutput("gosling_plot")
)


categories = c("Likely_benign", "Likely_pathogenic", "ambiguous")


track2_data <- track_data_gr(
    as_GRanges,chromosomeField = "seqnames", genomicFields = c("start", "end"))

ui <- fluidPage(
    use_gosling(clear_files = FALSE),
    goslingOutput("gosling_plot")
)

track_2 <- add_single_track(
    width = 800,
    height = 180,
    data = track2_data,
    mark = "text",
    x = visual_channel_x(
        field = "start", type = "genomic", axis = "bottom"
    ),
    xe = visual_channel_x(field = "end", type = "genomic"),
    y = visual_channel_y(
        field = "am_pathogenicity", type = "quantitative", axis = "right"
    ),
    color = visual_channel_color(
        field = "am_pathogenicity",
        type = "quantitative",
    ),
     tooltip = visual_channel_tooltips(
      visual_channel_tooltip(field = "REF", type = "nominal",
                             alt = "Reference"),
      visual_channel_tooltip(field = "ALT", type = "nominal",
                             alt = "Alternative / Mutation"),
      visual_channel_tooltip(
        field = "am_pathogenicity",
        type = "quantitative",
        alt = "AM_Pathogenicity Score",
        format = "0.2"
      ) ),
    text = list(field="ALT", type ="nominal"),
    size = list(value = 20)
)

composed_view2 <- compose_view(
    layout = "linear",
    xDomain = list(chromosome = "chr2",interval = c(46000, 46400)),
    tracks = track_2
    
)

arranged_view2 <- arrange_views(
    title = "Test track track : Q1W6H9 ",
    subtitle = "Gosling Example with text labels",
    views = composed_view2
)

server <- function(input, output, session) {
    output$gosling_plot <- renderGosling({
        gosling(
            component_id = "component_2",
            arranged_view2
        )
    })
}

shiny::shinyApp(ui, server)
```


```{r, multi-track}
library(shiny.gosling)
library(shiny)

ui <- fluidPage(
  use_gosling(clear_files = FALSE),
  goslingOutput("gosling_plot")
)

categories = c("likely_benign", "ambiguous","likely_pathogenic")

colormapping = c( "#029F73", "gray", "#CB3B8C")

get_range<-range(as_GRanges)

track3_data <- track_data_gr(
    as_GRanges,chromosomeField = "seqnames", genomicFields = c("start", "end"))

ui <- fluidPage(
    use_gosling(clear_files = FALSE),
    goslingOutput("gosling_plot")
)

track_3ref <- add_single_track(
    data = track3_data,
    mark = "rect",
    x = visual_channel_x(
        field = "start", type = "genomic", axis = "bottom"
    ),
    xe = visual_channel_x(field = "end", type = "genomic"),
    size = list(value = 50),
    stroke= "lightgrey",
    strokeWidth=list(value=1),
    opacity = list(value= 0.3)
)

track_3alt <- add_single_track(

    data = track3_data,
    mark = "point",
    x = visual_channel_x(
        field = "start", type = "genomic", axis = "bottom"
    ),
    xe = visual_channel_x(field = "end", type = "genomic"),
    y = visual_channel_y(field = "am_class", type = "nominal", axis = "right"),
    text = list(field="ALT", type ="nominal"),
    size = list(value = 5)
)

track_3overlay <- 
   
    
    # xDomain = list(chromosome = "chr2",interval = c(46000, 46400)),


composed_view3 <- compose_view(
    width = 800,
    height = 180,
    multi = TRUE,
    layout = "linear",
    xDomain = list(chromosome = as.character(get_range@seqnames@values),  interval = c(get_range@ranges@start, get_range@ranges@start+get_range@ranges@width)),
    alignment = "overlay",
    color = visual_channel_color(
        field = "am_class",
        type = "nominal",
        range = colormapping,
        legend = TRUE),
    tracks = add_multi_tracks(track_3ref,track_3alt)
    )

arranged_view3 <- arrange_views(
    title = "Test track track : Q1W6H9 ",
    subtitle = "stacked nucleotide example",
    views = composed_view3
)

server <- function(input, output, session) {
    output$gosling_plot <- renderGosling({
        gosling(
            component_id = "component_3",
            arranged_view3
        )
    })
}

shiny::shinyApp(ui, server)
```

```{r, baseplot}
library(shiny.gosling)
library(shiny)

ui <- fluidPage(
  use_gosling(clear_files = FALSE),
  goslingOutput("gosling_plot")
)

track4_data <- track_data_gr(
    as_GRanges,chromosomeField = "seqnames", genomicFields = c("start", "end"))

ui <- fluidPage(
    use_gosling(clear_files = FALSE),
    goslingOutput("gosling_plot")
)


bar <- add_single_track(
  mark = "bar",
  y = visual_channel_y(
    field = "count", type = "quantitative", axis = "none"
  )
)

text <- add_single_track(

  mark = "text",
  x = visual_channel_x(
    field = "start", type = "genomic"
  ),
  xe = visual_channel_x(
    field = "end", type = "genomic"
  ),
  size = 24,
  color = "white",
  visibility = list(list(
    operation = "less-than",
    measure = "width",
    threshold = "|xe-x|",
    transitionPadding = 30,
    target = "mark"
  ),
  list(
    operation = "LT",
    measure = "zoomLevel",
    threshold = 40,
    target = "track"
  ))
)

bar_x <- visual_channel_x(
  field = "position", type = "genomic"
)

bar_color <- visual_channel_color(
  field = "REF",
  type = "nominal",
  domain = c("A", "T", "G", "C"),
  legend = TRUE
)

bar_text <- visual_channel_text(
  field = "REF", type = "nominal"
)

bar_style <- default_track_styles(
  inlineLegend = TRUE
)

track4 <- add_single_track(
  title = "Sequence",
  alignment = "overlay",
  data = track4_data,
  tracks = add_multi_tracks(
    text, bar
  ),
  x = bar_x,
  color = bar_color,
  text = bar_text,
  style = bar_style,
  width = 800, height = 40
)

view_all <- compose_view(
  multi = TRUE,
  centerRadius = 0,
  xDomain = list(chromosome = "chr2",interval = c(46000, 46400)),
  linkingId = "detail",
  alignment = "stack",
  tracks = add_multi_tracks(
    track4
  )
)


arranged_view4 <- arrange_views(
    title = "Test track track : Q1W6H9 ",
    subtitle = "stacked nucleotide example",
    views = view_all
)

server <- function(input, output, session) {
    output$gosling_plot <- renderGosling({
        gosling(
            component_id = "component_4",
            arranged_view4
        )
    })
}

shiny::shinyApp(ui, server)
```





```{r}
library(shiny.gosling)
library(shiny)

as_GRanges <- as(gpos, "GRanges")


  # View 2 Track 3----
view2_track3_data <- track_data(
    url = "https://server.gosling-lang.org/api/v1/tileset_info/?d=sequence-multivec",
    type = "multivec",
    row = "base",
    column = "position",
    value = "count",
    categories = c("A", "T", "G", "C"),
    start = "start",
    end = "end"
  )



view2_track3a <- add_single_track(
    mark = "bar",
    y = visual_channel_y(
      field = "count", type = "quantitative", axis = "none"
    )
  )

view2_track3b <- add_single_track(
    dataTransform = track_data_transform(
      type = "filter",
      field = "count",
      oneOf = list(0),
      not = TRUE
    ),
    mark = "text",
    x = visual_channel_x(
      field = "start", type = "genomic"
    ),
    xe = visual_channel_x(
      field = "end", type = "genomic"
    ),
    size = 24,
    color = "white",
    visibility = list(
      list(
        operation = "less-than",
        measure = "width",
        threshold = "|xe-x|",
        transitionPadding = 30,
        target = "mark"
      ),
      list(
        operation = "LT",
        measure = "zoomLevel",
        threshold = 40,
        target = "track"
      )
    )
  )

  view2_track3_x <- visual_channel_x(
    field = "position", type = "genomic"
  )

  view2_track3_color <- visual_channel_color(
    field = "base",
    type = "nominal",
    domain = c("A", "T", "G", "C"),
    legend = TRUE
  )

  view2_track3_text <- visual_channel_text(
    field = "base", type = "nominal"
  )

  view2_track3_style <- default_track_styles(
    inlineLegend = TRUE
  )

  view2_track3 <- add_single_track(
    title = "NC_045512.2 Sequence",
    alignment = "overlay",
    data = view2_track3_data,
    tracks = add_multi_tracks(
      view2_track3a, view2_track3b
    ),
    x = view2_track3_x,
    color = view2_track3_color,
    text = view2_track3_text,
    style = view2_track3_style,
    width = 800, height = 40
  )

  view2 <- compose_view(
    multi = TRUE,
    centerRadius = 0,
    xDomain = list(chromosome = "chr12"),
    linkingId = "detail",
    alignment = "stack",
    tracks = add_multi_tracks(
      view2_track3
    )
  )

  combined_view <- arrange_views(
    title = "SARS-CoV-2",
    subtitle = "Data Source: WashU Virus Genome Browser, NCBI, GISAID",
    assembly = list(list("NC_045512.2", 29903)),
    layout = "linear",
    spacing = 50,
    views = list(view2),
    listify = FALSE
  )

  ui <- fluidPage(
    use_gosling(),
    fluidRow(
      column(6, goslingOutput("gosling_plot"))
    )
  )


  server <- function(input, output, session) {
    output$gosling_plot <- renderGosling({
      gosling(
        component_id = "sars_cov2",
        combined_view
      )
    })
  }

  shinyApp(ui, server)

```


```{r}
track1_data <- track_data(
  url = "https://server.gosling-lang.org/api/v1/tileset_info/?d=NC_045512_2-multivec",
  type = "multivec",
  row = "base",
  column = "position",
  value = "count",
  categories = c("A", "T", "G", "C"),
  start = "start",
  end = "end"
)

track1 <- add_single_track(
  mark = "bar",
  y = visual_channel_y(
    field = "count", type = "quantitative", axis = "none"
  )
  
)


track1_x <- visual_channel_x(
  field = "position", type = "genomic"
)

track1_color <- visual_channel_color(
  field = "base",
  type = "nominal",
  domain = c("A", "T", "G", "C"),
  legend = TRUE
)

track1_text <- visual_channel_text(
  field = "base", type = "nominal"
)

track1_style <- default_track_styles(
  inlineLegend = TRUE
)


composed_view <- compose_view(
    layout = "linear",
    tracks = track_1
)

arranged_view <- arrange_views(
    title = "Basic bar track : hg38",
    subtitle = "Gosling Example",
    views = composed_view
)

server <- function(input, output, session) {
    output$gosling_plot <- renderGosling({
        gosling(
            component_id = "component_1",
            arranged_view
        )
    })
}

shiny::shinyApp(ui, server)
```


# Finally

Remember to disconnect and shutdown all managed DuckDB connections.

```{r db_disconnect_all}
db_disconnect_all()
```

Database connections that are not closed correctly trigger warning
messages.

# Session information {.unnumbered}

```{r sessionInfo}
sessionInfo()
```
